internal/providers/azure/azure.go

```diff
@@ -56,15 +57,38 @@ const (
 	CDS_FSTYPE_UDF = "udf"
 )
 
-// FetchConfig wraps FetchOvfDevice to implement the platform.NewFetcher interface.
+var (
+	imdsURL = url.URL{
+		Scheme: "http",
+		Host:   "169.254.169.254",
+		Path:   "metadata/instance/compute/userData?api-version=2021-01-01&format=text",
+	}
+)
+
+// FetchConfig wraps FetchFromAzureMetadata to implement the platform.NewFetcher interface.
 func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
-	return FetchFromOvfDevice(f, []string{CDS_FSTYPE_UDF})
+	return FetchFromAzureMetadata(f, []string{CDS_FSTYPE_UDF})
 }
 
-// FetchFromOvfDevice has the return signature of platform.NewFetcher. It is
+// FetchFromAzureMetadata has the return signature of platform.NewFetcher. It is
 // wrapped by this and AzureStack packages.
-func FetchFromOvfDevice(f *resource.Fetcher, ovfFsTypes []string) (types.Config, report.Report, error) {
+// It first tries to fetch user-data from Azure IMDS and in case of failures, it fallbacks to custom-data.
+func FetchFromAzureMetadata(f *resource.Fetcher, ovfFsTypes []string) (types.Config, report.Report, error) {
+	// fetch-offline is not supported since we first try to get config from Azure IMDS.
+	// this config fetching can only run in fetch stage.
+	if f.Offline {
+		return types.Config{}, report.Report{}, resource.ErrNeedNet
+	}
+
 	logger := f.Logger
+
+	// we first try to config from IMDS, in case of failure we fallback on the custom data.
+	data, err := f.FetchToBuffer(imdsURL, resource.FetchOptions{Compression: "base64"})
+	if err == nil {
+		return util.ParseConfig(logger, data)
```

Comment:
For user debugging purposes, let's log at `Info` level if we've read the config from userdata and if we've read it from custom data.